README
$Id: README,v 1.7 2007/01/16 11:16:15 mike Exp $

The Net::Z3950::ZOOM, ZOOM and Net::Z3950 modules
=================================================

This distribution contains three Perl modules for the price of one.
They all provide facilities for building information retrieval clients
using the standard Z39.50 and SRW/U protocols, but do so using
different APIs.

- If you are new to this distribution, then you should use the ZOOM
  API, and ignore the others.  It is the cleanest, most elegant and
  intuitive, and most closely follows the letter as well as the spirit
  of the Abstract ZOOM API as specified at http://zoom.z3950.org/api/

- If you have used the old Net::Z3950 module and have to maintain an
  application that calls that API, then you will want to use the
  Net::Z3950 classes provided in this distribution, which provide an
  API compatible with the old module's implemented on top of the new
  ZOOM code.

- You should definitely not use the Net::Z3950::ZOOM API, which is not
  object-oriented, and instead provides the thinnest possible layer on
  top of the ZOOM-C functions in the YAZ toolkit.  This API exists
  only in order to have ZOOM API built on top of it.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


DEBIAN PACKAGES

To build Debian packages issue:

   dh-make-perl --build


DEPENDENCIES

This module requires these other modules and libraries:

- The YAZ toolkit for Z39.50 and SRW/U communication.  This is
  available as a package on several platform -- for example, Debian
  GNU/Linux supports "apt-get install yaz".  For others, you will need
  to download and build the source-code, which is much more
  straightforward that you probably expect.  You can get it from
  http://indexdata.com/yaz/

  NOTE THAT THE ZOOM-Perl MODULE ABSOLUTELY REQUIRES RELEASE 2.0.11 OR
  BETTER OF THE YAZ TOOLKIT.  You need version 2.1.17 or better if you
  want to run clever asynchronous programs that use the END event,
  which did not exist prior to that release.

COPYRIGHT AND LICENCE

Copyright (C) 2005-2007 by Index Data.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.




Changes
$Id: Changes,v 1.69 2007/09/21 16:52:26 mike Exp $

Revision history for Perl extension Net::Z3950::ZOOM.

1.20  Fri Sep 21 17:51:55 BST 2007
	- Add new sample program, samples/zoom/zoomscan.pl
	- samples/zoom/zoomtst1.pl now accepts arbitrary options at
	  the end of the command-line, so that (for example) you can
	  specify a non-USMARC record syntax.
	- Add new functions Net::Z3950::ZOOM::diag_srw_str() and
	  ZOOM::diag_srw_str() to map SRW diagnostic codes into
	  human-readable messages.
	- ZOOM::Exception() constructor uses diag_srw_str() to fill in
	  the error message, if it is undefined, for SRW errors.
	- The render() method of ZOOM::Exception(), and therefore the
	  overloaded stringifying operator, deals with exceptions that
	  have no message().
	- Corrected the scan parts of the test-suite that use CQL so
	  that they explicitly perform client-side CQL compilation and
	  send the resulting Type-1 query.  Previously, these tests
	  were (accidentally) requesting that CQL be sent as-is to the
	  server, but this mistake was being masked by an error in
	  older versions of ZOOM-C that compiled client-side anyway!

1.19  Mon Jul  9 14:09:31 BST 2007
	- Add $conn->exception() method.
	- Set version number for Net::Z3950 to 0.99 -- less than the
	  1.0 version number that will be used when it successfully
	  emulates the old Net::Z3950 module, but superseding the last
	  release, 0.51, of that module.
	- Include samples/zoom/zdump.pl in distribution.
	- Add "irspy" namespace to samples/zoom/zselect

1.18  Mon Feb 26 14:57:48 GMT 2007
	- When a connection is in asynchronous mode, failing
	  operations (search, scan, etc.) no longer throw exceptions.
	  This is because the event-loop in asychronous applications
	  needs to check for errors anyway, in case the server has
	  returned high-level diagnostics such as "unsupported use
	  attribute".  So this check within the event-loop is now the
	  only place where error-checking need be done.
	- Add new public ZOOM::Connection() function, check(), which
	  throws an exception if an error is pending (even if the
	  connection is asynchronous).
	- New YAZ logging level, "zoom_check", notes when errors are
	  detected and either left unreported (asynchronous
	  connections) or thrown as exception (synchronous).

1.17  Thu Feb 22 20:38:45 GMT 2007
	- Change Net::Z3950::ZOOM::event() to allocate its array of
	  connection pointers dynamically, so there is no longer an
	  arbitrary limit of 100.
	- New connection-level option "_check_debug" emits messages on
	  standard error when checking a connection find an error that
	  is about to result in throwing an exception.  There is no
	  reason for you ever to use this, though -- pretend you don't
	  know about it.

1.16  Tue Jan 16 11:18:34 GMT 2007
	- Require version 2.1.41 or better of YAZ, so that Init
	  Response options are available using
	  $conn->option("init_opt_sort") etc.
	- Include samples/zoom/zhello.pl to exercise these options.
	- Add ZOOM_connection_peek_event() and $conn->peek_event().
	- Require YAZ version 2.1.45, which is the first to provide
	  ZOOM_connection_peek_event().

1.15  Fri Dec  1 14:17:49 GMT 2006
	- ZOOM::Error::INIT is renamed to ZOOM::Error::ZINIT, since
	  INIT is a reserved word in Perl.  In version 1.14 and
	  earlier, "require ZOOM" would fail, saying "Too late to run
	  INIT block", and "use ZOOM" would result in
	  ZOOM::Error::INIT having a value equal to its name (why?!)
	  rather than, as intended, Net::Z3950::ZOOM::ERROR_INIT.
	- Belatedly added documentation for $rec->error() and
	  $rec->exception().

1.14  Tue Nov 28 17:33:07 GMT 2006
	- Add $record->error() to return non-surrogate diagnostics,
	  and $record->exception() to return the same information
	  wrapped in a ZOOM::Exception object.
	- Requires YAZ 2.1.40, which provides ZOOM_record_error().
	- $conn->error_x() now returns the error-code when called in
	  scalar context, rather than the diagnostic set name.

1.13  Sat Nov  4 16:47:00 GMT 2006
	- ZOOM::Connection::create() may now take either a single
	  argument, which is a ZOOM::Options object, or any even
	  number of argument (including zero), which are key => value
	  pairs of options to set.
	- ZOOM::Connection::new(), having been refactored to use
	  create(), can now also take a ZOOM::Options argument.
	- Documentation now includes information on specifying SRW,
	  SRU-over-GET and SRU-over-POST connections.
	- ZOOM::ResultSet::record() throws a diagnostic if asked to
	  retrieve records out of range for the result-set.
	- All tests against indexdata.com are now explicitly against
	  z3950.indexdata.com, as the identity of the former name has
	  recently changed.

1.12  (NEVER PUBLICLY RELEASED; these changes are included in 1.13)
	- Include the useful example program samples/zoom/zselect,
	  which is installed as part of "make install".
	- Add ZOOM_connection_is_idle() and $conn->is_idle().
	- Require YAZ version 2.1.35, which is the first to provide
	  ZOOM_connection_is_idle().
	- Fix bug in ZOOM::Connection class code that made the
	  difficult to subclass -- when called as
	  $class->SUPER::new(@_), the constructor blessed the new
	  object into ZOOM::Connection rather than info the subclass.

1.11  Thu Aug 31 16:47:53 BST 2006
	- Require YAZ version 2.1.17, which has important ZOOM-C
	  bug-fixes.
	- Do not use "-Wdeclaration-after-statement" compiler option,
	  since this supported only by GCC (and only recent versions).

1.10  Thu Jun 15 16:42:47 BST 2006
	- No functional changes, but use Adam's modified API to
	  ZOOM_query_ccl2rpn().  This incompatible change means that
	  RELEASE 1.09 WILL NOT BUILD against any recent YAZ.

1.09  Tue Jun 13 17:44:43 2006
	- Add new function Net::Z3950::ZOOM::query_ccl2rpn(), for
	  client-side CCL compilation.
	- Add new ZOOM::Query::CCL2RPN class, encapsulating CCL
	  compiler functionality as a Query subclass.  This allows
	  client-side CCL to be used for both search and scan.

lib/Net/Z3950.pm
# $Id: Z3950.pm,v 1.2 2007/02/26 15:58:33 mike Exp $

use strict;
use warnings;
use Net::Z3950::ZOOM;

our $VERSION = '0.99';	# Supersedes "proper" Net::Z3950 v0.51


# Member naming convention is the same as in ../ZOOM.pm

# ----------------------------------------------------------------------------
# Enumerations are copied from the old Net::Z3950 module.
# It's not entirely clear yet which of these we actually need, for for
# now I am commenting them all out, and faulting in the ones we need.
# ----------------------------------------------------------------------------

# Define the operation-code enumeration.  The values here are chosen
# to be in a distinctive range (i.e. 3950 plus a small integer) so
# that if they are misused in another context, they're easy to spot.
package Net::Z3950::Op;
sub Error    { 3951 }
sub Init     { 3952 }
sub Search   { 3953 }
sub Get      { 3954 }
sub DeleteRS { 3955 }
sub Scan     { 3956 }
package Net::Z3950;


## Define the record-syntax enumeration.  These values must be kept
## synchronised with the values implied by the oid_value enumeration in
## the header file "yaz/oid.h"
#package Net::Z3950::RecordSyntax;
#sub UNIMARC    { 16 }
#sub INTERMARC  { 17 }
#sub CCF        { 18 }
#sub USMARC     { 19 }
#sub UKMARC     { 20 }
#sub NORMARC    { 21 }
#sub LIBRISMARC { 22 }
#sub DANMARC    { 23 }
#sub FINMARC    { 24 }
#sub MAB        { 25 }
#sub CANMARC    { 26 }
#sub SBN        { 27 }
#sub PICAMARC   { 28 }
#sub AUSMARC    { 29 }
#sub IBERMARC   { 30 }
#sub CATMARC    { 31 }
#sub MALMARC    { 32 }
#sub EXPLAIN    { 33 }
#sub SUTRS      { 34 }
#sub OPAC       { 35 }
#sub SUMMARY    { 36 }
#sub GRS0       { 37 }
#sub GRS1       { 38 }
#sub EXTENDED   { 39 }
#sub TEXT_HTML  { 70 }
#sub XML        { 80 }
#sub TEXT_XML   { 80 }
#sub APPLICATION_XML { 81 }
#
#use vars '%map';
## Maps record-syntax name strings to enumeration members
#%map = (UNIMARC => UNIMARC,
#	INTERMARC => INTERMARC,
#	CCF => CCF,
#	USMARC => USMARC,
#	UKMARC => UKMARC,
#	NORMARC => NORMARC,
#	LIBRISMARC => LIBRISMARC,
#	DANMARC => DANMARC,
#	FINMARC => FINMARC,
#	MAB => MAB,
#	CANMARC => CANMARC,
#	SBN => SBN,
#	PICAMARC => PICAMARC,
#	AUSMARC => AUSMARC,
#	IBERMARC => IBERMARC,
#	CATMARC => CATMARC,
#	MALMARC => MALMARC,
#	EXPLAIN => EXPLAIN,
#	SUTRS => SUTRS,
#	OPAC => OPAC,
#	SUMMARY => SUMMARY,
#	GRS0 => GRS0,
#	GRS1 => GRS1,
#	EXTENDED => EXTENDED,
#	TEXT_HTML => TEXT_HTML,
#	XML => XML,
#	TEXT_XML => TEXT_XML,
#	APPLICATION_XML => APPLICATION_XML,
#	);
#package Net::Z3950;
#
#
## Define the reason-for-decodeAPDU()-failure enumeration.  This must
## be kept synchronised with the values #defined in "yazwrap/yazwrap.h"
#package Net::Z3950::Reason;
#sub EOF        { 23951 }	# read EOF from connection (server gone)
#sub Incomplete { 23952 }	# read bytes, but not yet a whole APDU
#sub Malformed  { 23953 }	# couldn't decode APDU (malformed)
#sub BadAPDU    { 23954 }	# APDU was well-formed but unrecognised
#sub Error      { 23955 }	# some other error (consult errno)
#package Net::Z3950;
#
#
## Define the query-type enumeration.  This must be kept synchronised
## with the values #defined in "yazwrap/yazwrap.h"
#package Net::Z3950::QueryType;
#sub Prefix  { 39501 }		# Yaz's "@attr"-ish forward-Polish notation
#sub CCL     { 39502 }		# Send CCL string to server ``as is''
#sub CCL2RPN { 39503 }		# Convert CCL to RPN (type-1) locally
#sub CQL     { 39504 }		# Send CQL string to server ``as is''
#package Net::Z3950;
#
#
## Define the result-set-status enumeration, used by the
## `resultSetStatus' field in the Net::Z3950::APDU::SearchResponse
## class in cases where `searchStatus' is false (indicating failure).
## This must be kept synchronised with the ASN.1 for the structure
## described in section 3.2.2.1.11 of the Z39.50 standard itself.
#package Net::Z3950::ResultSetStatus;
#sub Subset  { 1 }
#sub Interim { 2 }
#sub None    { 3 }
#package Net::Z3950;
#
#
## Define the present-status enumeration, used by the `presentStatus'
## field in the Net::Z3950::APDU::SearchResponse class in cases where
## `searchStatus' is true (indicating success).  This must be kept
## synchronised with the ASN.1 for the structure described in section
## 3.2.2.1.11 of the Z39.50 standard itself.
#package Net::Z3950::PresentStatus;
#sub Success  { 0 }
#sub Partial1 { 1 }
#sub Partial2 { 2 }
#sub Partial3 { 3 }
#sub Partial4 { 4 }
#sub Failure  { 5 }
#package Net::Z3950;
#
#
## Define the scan-status enumeration, used by the `scanStatus'
## field in the Net::Z3950::APDU::ScanResponse class.  This must be
## kept synchronised with the ASN.1 for the structure described in
## section 3.2.8.1.6 of the Z39.50 standard itself.
#package Net::Z3950::ScanStatus;
#sub Success  { 0 }
#sub Partial1 { 1 }
#sub Partial2 { 2 }
#sub Partial3 { 3 }
#sub Partial4 { 4 }
#sub Partial5 { 5 }
#sub Failure  { 6 }
#package Net::Z3950;

# ----------------------------------------------------------------------------

package Net::Z3950;

sub errstr {
    my($errcode) = @_;
    # This is not 100% compatible, because it will translate
    # ZOOM-level errors as well as BIB-1 diagnostic codes.
    return Net::Z3950::ZOOM::diag_str($errcode)
}

sub opstr {
    my($op) = @_;
    return "error" if $op == Net::Z3950::Op::Error;
    return "init" if $op == Net::Z3950::Op::Init;
    return "search" if $op == Net::Z3950::Op::Search;
    return "get" if $op == Net::Z3950::Op::Get;
    return "deleteRS" if $op == Net::Z3950::Op::DeleteRS;
    return "scan" if $op == Net::Z3950::Op::Scan;
    return "unknown op " . $op;
}


# ----------------------------------------------------------------------------

package Net::Z3950::Manager;
sub new { Net::Z3950::Connection->new() }


# ----------------------------------------------------------------------------

package Net::Z3950::Connection;

sub new {
    die "The Net::Z3950::ZOOM distribution does not yet support the Net::Z3950 'Classic' API.  A subsequent version will do so; until then, please continue using Net::Z3950 itself if you need its API."
}


1;
